rust/routee-compass/src/plugin/input/default/edge_rtree/edge_rtree_input_plugin.rs

```diff
@@ -0,0 +1,162 @@
+use super::edge_rtree_record::EdgeRtreeRecord;
+use crate::{
+    app::compass::config::{
+        compass_configuration_error::CompassConfigurationError,
+        config_json_extension::ConfigJsonExtensions,
+    },
+    plugin::{
+        input::{input_json_extensions::InputJsonExtensions, input_plugin::InputPlugin},
+        plugin_error::PluginError,
+    },
+};
+use geo_types::Coord;
+use routee_compass_core::{
+    model::road_network::edge_id::EdgeId,
+    util::{
+        fs::{read_decoders, read_utils},
+        geo::geo_io_utils::read_linestring_text_file,
+        unit::{as_f64::AsF64, Distance, DistanceUnit, BASE_DISTANCE_UNIT},
+    },
+};
+use rstar::RTree;
+use std::{collections::HashSet, path::Path};
+
+pub struct EdgeRtreeInputPlugin {
+    pub rtree: RTree<EdgeRtreeRecord>,
+    pub tolerance: Option<(Distance, DistanceUnit)>,
+}
+
+impl InputPlugin for EdgeRtreeInputPlugin {
+    /// finds the nearest edge ids to the user-provided origin and destination coordinates.
+    /// optionally restricts the search to a subset of road classes tagged by the user.
+    fn process(&self, query: &serde_json::Value) -> Result<Vec<serde_json::Value>, PluginError> {
+        let road_classes_key = String::from("road_classes");
+        let road_classes = query
+            .get_config_serde_optional::<HashSet<String>>(road_classes_key, String::from(""))
+            .map_err(|e| {
+                PluginError::InputError(format!("unable to deserialize as array: {}", e))
+            })?;
+        let src_coord = query.get_origin_coordinate()?;
+        let dst_coord_option = query.get_destination_coordinate()?;
+
+        let source_edge_id = search(&self.rtree, src_coord, &road_classes, self.tolerance)
+            .ok_or(matching_error(&src_coord, self.tolerance))?;
+        let destination_edge_id_option = match dst_coord_option {
+            None => Ok(None),
+            Some(dst_coord) => search(&self.rtree, dst_coord, &road_classes, self.tolerance)
+                .map(Some)
+                .ok_or(matching_error(&dst_coord, self.tolerance)),
+        }?;
+
+        let mut updated = query.clone();
+        updated.add_origin_edge(source_edge_id)?;
+        match destination_edge_id_option {
+            None => {}
+            Some(destination_edge_id) => {
+                updated.add_destination_edge(destination_edge_id)?;
+            }
+        }
+
+        Ok(vec![updated])
+    }
+}
+
+impl EdgeRtreeInputPlugin {
+    pub fn new(
+        road_class_file: &Path,
+        linestring_file: &Path,
+        tolerance_distance: Option<Distance>,
+        distance_unit: Option<DistanceUnit>,
+    ) -> Result<Self, CompassConfigurationError> {
+        let road_class_lookup: Box<[String]> =
```

Comment:
It could be nice to eventually allow arbitrary frontier models here instead of just road class. For example, if we wanted to apply this to heavy duty trucks where there might be restrictions based on vehicle height/weight/length.

Reply:
yup. i have another thought from doing this. the user should be able to set these requirements at _query time_. the FrontierModel should be 1) parameterized just like TraversalModelService, or 2) should just be rolled in as a sub-component of a TraversalModel and with some kind of build behavior based on query flags/parameters.